sound/soc/sof/ipc4-pcm.c

```diff
@@ -403,10 +404,56 @@ static int sof_ipc4_pcm_close(struct snd_soc_component *component,
 	return 0;
 }
 
+static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
+					    struct snd_pcm_substream *substream)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct sof_ipc4_timestamp_info *time_info;
+	struct sof_ipc4_llp_reading_slot llp;
+	struct snd_sof_pcm_stream *stream;
+	struct snd_sof_pcm *spcm;
+	u64 link_pos;
+
+	spcm = snd_sof_find_spcm_dai(component, rtd);
+	if (!spcm)
+		return 0;
+
+	stream = &spcm->stream[substream->stream];
+	time_info = stream->time_info;
+
+	if (time_info->stream_start_offset == SOF_IPC4_INVALID_STREAM_POSITION) {
+		struct sof_ipc4_pipeline_registers ppl_reg;
+
+		sof_mailbox_read(sdev, sdev->fw_info_box.offset + time_info->ppl_reg_offset,
+				 &ppl_reg, sizeof(ppl_reg));
+
+		if (ppl_reg.stream_start_offset == SOF_IPC4_INVALID_STREAM_POSITION) {
+			dev_err(sdev->dev, "failed to get stream start offset\n");
+			return 0;
+		}
+
+		time_info->stream_start_offset = ppl_reg.stream_start_offset;
+	}
+
+	if (time_info->llp_offset)
+		sof_mailbox_read(sdev, time_info->llp_offset, &llp, sizeof(llp));
+
+	sof_ops(sdev)->pcm_pointer(sdev, substream);
```

Comment:
@kv2019i  please check the hw delay in Linux document. In my opinion it is not the ptr from last pointer. And legacy hda also supports delay. it doesn't use last ptr point. please check azx_get_delay_from_lpib and its usage
````c
--------------------------------------------------------------> time
  ^               ^              ^                ^           ^
  |               |              |                |           |
 analog         link            dma              app       FullBuffer
 time           time           time              time        time
  |               |              |                |           |
  |< codec delay >|<--hw delay-->|<queued samples>|<---avail->|
  |<----------------- delay---------------------->|           |
                                 |<----ring buffer length---->|
````

Reply:
@RanderWang the azx_get_delay_from_lpib() uses the DPIB position provided earlier. It does NOT recheck for the DPIB position once more. That would really make the delay lower than it is, if the driver reported the 'old' DPIB read and the delay value based on the 'new' DPIB read.

Of course this is not ideal since there's no instant snapshot of DPIB and LPIB so there's got to be variations in the delay. Still we have to be consistent in the arithmetic with the baseline that's use for DPIB.